pkg/envoy/registry/services_test.go

```diff
@@ -250,3 +170,1173 @@ var _ = Describe("Test Proxy-Service mapping", func() {
+			kubeController := k8s.NewMockController(ctrl)
+			kubeController.EXPECT().ListPods().Return(test.existingPods).Times(1)
+
+			actual := listPodsForService(test.service, kubeController)
+			tassert.Equal(t, test.expected, actual)
+		})
+	}
+}
+
+func TestGetCertCommonNameForPod(t *testing.T) {
+	tests := []struct {
+		name      string
+		pod       v1.Pod
+		shouldErr bool
+	}{
+		{
+			name: "valid CN",
+			pod: v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "ns",
+					Labels: map[string]string{
+						constants.EnvoyUniqueIDLabelName: uuid.New().String(),
+					},
+				},
+				Spec: v1.PodSpec{
+					ServiceAccountName: "svcacc",
+				},
+			},
+			shouldErr: false,
+		},
+		{
+			name: "invalid UID",
+			pod: v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "ns",
+					Labels: map[string]string{
+						constants.EnvoyUniqueIDLabelName: uuid.New().String() + "-not-valid",
+					},
+				},
+				Spec: v1.PodSpec{
+					ServiceAccountName: "svcacc",
+				},
+			},
+			shouldErr: true,
+		},
+		{
+			name: "no UID",
+			pod: v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "ns",
+					Labels:    map[string]string{},
+				},
+				Spec: v1.PodSpec{
+					ServiceAccountName: "svcacc",
+				},
+			},
+			shouldErr: true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			assert := tassert.New(t)
+			cn, err := getCertCommonNameForPod(test.pod)
+			if test.shouldErr {
+				assert.Empty(cn)
+				assert.Error(err)
+			} else {
+				assert.NotEmpty(cn)
+				assert.NoError(err)
+			}
+		})
+	}
+}
+
+func TestKubeProxyServiceMapperRace(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	kubeController := k8s.NewMockController(mockCtrl)
+
+	k := NewKubeProxyServiceMapper(kubeController)
+
+	stop := make(chan struct{})
+	defer close(stop)
+
+	proxyUUID := uuid.New()
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: map[string]string{
+				constants.EnvoyUniqueIDLabelName: proxyUUID.String(),
+				"app":                            "my-app",
+			},
+		},
+	}
+	svc := &v1.Service{
+		Spec: v1.ServiceSpec{
+			Selector: map[string]string{
+				"app": "my-app",
+			},
+		},
+	}
+	cn := envoy.NewCertCommonNameWithProxyID(proxyUUID, pod.Spec.ServiceAccountName, pod.Namespace)
+	proxy := envoy.NewProxy(cn, "", nil)
+
+	doWrites := func(stop <-chan struct{}) {
+		for {
+			select {
+			case <-stop:
+				return
+			default:
+				kubeController.EXPECT().ListServices().Return(nil).Times(1)
+				k.handlePodUpdate(pod)
+				kubeController.EXPECT().ListPods().Return([]*v1.Pod{pod}).Times(1)
+				k.handleServiceUpdate(svc)
+				k.handleServiceDelete(svc)
+				k.handlePodDelete(pod)
+			}
+		}
+	}
+
+	doReads := func(stop <-chan struct{}) {
+		for {
+			select {
+			case <-stop:
+				return
+			default:
+				_, _ = k.ListProxyServices(proxy)
+			}
+		}
+	}
+
+	go doWrites(stop)
+	go doReads(stop)
+	go doReads(stop)
+
+	time.Sleep(250 * time.Millisecond)
```

Comment:
@draychev, @shashankram I added this test to verify that this is resilient against concurrent reads and writes, but I'm not a huge fan of the `time.Sleep()` here. Can you think of a better way to lay this out?

Reply:
Why not just rely on the Go testing's race detector instead? This will need additional signaling logic to avoid the `sleep`.

Reply:
The test does rely on `go test -race`, but the test needs to run for some period of time for the race condition to be detected. The other tests do not trigger any race conditions.